testkit/tests/runtime.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use exonum::{
+    blockchain::InstanceConfig,
+    crypto::{PublicKey, SecretKey},
+    node::ApiSender,
+    proto::Any,
+    runtime::dispatcher::{Dispatcher, DispatcherSender},
+    runtime::{
+        ArtifactId, ArtifactProtobufSpec, CallInfo, ExecutionContext, ExecutionError,
+        InstanceDescriptor, InstanceSpec, Runtime, StateHashAggregator,
+    },
+};
+use exonum_merkledb::{Fork, Snapshot};
+use exonum_testkit::{runtime::RuntimeFactory, TestKitBuilder};
+use futures::{Future, IntoFuture};
+use std::{cell::RefCell, sync::Arc};
+
+// Tracks parts of runtime's state that we're interested in.
+#[derive(Debug)]
+struct RuntimeState {
+    deployed_artifact: ArtifactId,
+    deploy_spec: Any,
+    config_params: Any,
+}
+
+// Main purpose is to track and make some assertions on state of runtime.
+#[derive(Debug)]
+struct RuntimeTester {
+    state: RefCell<RuntimeState>,
+}
+
+impl RuntimeTester {
+    fn deploy_artifact(&self, artifact: ArtifactId, deploy_spec: Any) {
+        let mut state = self.state.borrow_mut();
+        state.deployed_artifact = artifact;
+        state.deploy_spec = deploy_spec;
+    }
+
+    fn configure_service(&self, config_params: Any) {
+        let mut state = self.state.borrow_mut();
+        state.config_params = config_params;
+    }
+
+    fn is_artifact_deployed(&self, artifact_id: &ArtifactId) -> bool {
+        let state = self.state.borrow();
+        state.deployed_artifact == *artifact_id
+    }
+
+    fn assert_artifact_deployed(&self, artifact_id: ArtifactId, deploy_spec: Any) {
+        let state = self.state.borrow();
+        assert_eq!(state.deployed_artifact, artifact_id);
+        assert_eq!(state.deploy_spec, deploy_spec);
+    }
+
+    fn assert_config_params_passed(&self, config_params: Any) {
+        let state = self.state.borrow();
+        assert_eq!(state.config_params, config_params)
+    }
+}
+
+impl Default for RuntimeTester {
```

Comment:
nit: Why do you need manual `Default` implementation? `derive` would work

Reply:
`derive` doesn't work because of `ArtifactId` that doesn't have `Default` implementation.

Reply:
So, maybe you should create more meaningfull artifact by default.

Reply:
I really don't think there should be a default implementation of `ArtifactId`, so this code is okay (though a bit verbose)